.github/workflows/ci.yml

```diff
@@ -44,3 +44,43 @@ jobs:
 
     - name: Run Tests
       run: dotnet test test\OpenFeature.Tests\ --configuration Release --logger GitHubActions
+
+  packaging:
+    needs:
+    - unit-tests-linux
+    - unit-tests-windows
+
+    permissions:
+      contents: read
+      packages: write
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v4
+
+    - name: Setup .NET SDK
+      uses: actions/setup-dotnet@v4
+      with:
+        dotnet-version: |
+          6.0.x
+          7.0.x
+
+    - name: Restore
+      run: dotnet restore
+
+    - name: Pack NuGet packages (CI versions)
+      if: startsWith(github.ref, 'refs/heads/')
```

Comment:
I'm not sure how this will work, because we seem to have this job setup to run [only on `main`](https://github.com/open-feature/dotnet-sdk/pull/173/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR4-R11).

I wonder if a neater solution would be to do something like we do for Java, and just publish every merge to main to Github.

That wouldn't get us testing for branches, but I think it might be good enough?

Reply:
I always have to refer to the docs when messing with the `on` syntax _(hence that oops you caught lol)_, but `on.pull_request: [ main ]` results in this workflow running for PRs _targeting_ `main` ___(emphasis added)___:

> Runs your workflow when activity on a pull request in the workflow's repository occurs. For example, if no activity types are specified, the workflow ___runs when___ a pull request is opened or reopened or ___when the head branch of the pull request is updated___.
>
> [...]
> 
> For example, this workflow will run when someone opens a pull request that targets a branch whose name starts with `releases/`:
> 
> ```yaml
> on:
>   pull_request:
>     types:
>       - opened
>     branches:
>       - 'releases/**'
> ```
>
> _via https://docs.github.com/actions/using-workflows/events-that-trigger-workflows#pull_request_

which seems to be true based on what I'm seeing on this PR (after fixing the `pull_request{_target => }` issue):

![image](https://github.com/open-feature/dotnet-sdk/assets/21338699/c9e3f8bd-8590-4617-b50d-a6e6fcd4e5f0)

Reply:
> [...]
>
> I wonder if a neater solution would be to do something like we do for Java, and just publish every merge to main to Github.
>
> That wouldn't get us testing for branches, but I think it might be good enough?

The workflow, as-is, _will_ publish a `x.y.z-ci.$(date -u +%Y%m%dT%H%M%S)+sha.${GITHUB_SHA:0:9}` package for every commit on `main`, so we're already good on that.

But the sticky part is seeing if we can get _fork_ PRs to publish the `x.y.z-pr.$(date -u +%Y%m%dT%H%M%S)+sha.${GITHUB_SHA:0:9}` like I'd hoped we'd be able to.

I use a workflow like this for internal packages all the time, so pretty sure its the `pr from fork` part that's causing issues.

Let me try one more tweak to try getting it to publish to the _fork's package feed_, and if that doesn't work, I'll set this step to skip for PRs until we can figure out another solution (e.g. uploading the nuget package as an artifact, but its always a PITA to use local packages, so was hoping to find a nicer solution).

Reply:
Right OK, I confused myself on that one :sweat_smile: 

So, if I'm understanding correctly, your intent is to trigger this for any branch or tag, or any PR to main? I wonder if that's a bit excessive?

Reply:
As you can see in this case, it will run (and fail) with every PR from a fork due to missing secrets.

Reply:
I wonder if it's not better to instead only run this for a specific tag pattern or branch, or maybe on every push to main?

Reply:
Okay last tweak to try and get this working:

```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index f9aa524..5649f08 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -81,6 +81,13 @@ jobs:
       if: startsWith(github.ref, 'refs/tags/v')
       run: dotnet pack --no-restore

-    - name: Publish NuGet packages
-      if: startsWith(github.ref, 'refs/heads/') || startsWith(github.ref, 'refs/pull/') || startsWith(github.ref, 'refs/tags/v')
+    - name: Publish NuGet packages (base)
+      if: github.event.pull_request.head.repo.fork == false
       run: dotnet nuget push "src/**/*.nupkg" --api-key "${{ secrets.GITHUB_TOKEN }}" --source https://nuget.pkg.github.com/open-feature/index.json
+
+    - name: Publish NuGet packages (fork)
+      if: github.event.pull_request.head.repo.fork == true
+      uses: actions/upload-artifact@v4.2.0
+      with:
+        name: nupkgs
+        path: src/**/*.nupkg
```

> So, if I'm understanding correctly, your intent is to trigger this for any branch or tag, or any PR to main?

Almost, my intent was to have this run for every push to `main` (excluding `**.md`-only changes), which is what the `ci.yml` workflow was already setup to do.

We could do it for tags too (PR includes support for that), but we would have to add `on.push.tags: [ v* ]` which I have not included here (yet?).


> I wonder if that's a bit excessive?

Personally, I don't think its excessive, but I could imagine it someday _maybe_ becoming excessive, but there are ways we could mitigate if/when that occurs (e.g. changing the `main` packages to run on a cron schedule conditional on changes having landed since the last run).

As for the _volume_ of packages this _could_ add to our package feed, we can setup a workflow to prune the CI packages, keeping only the most recent N versions, if it becomes an issue (or proactively, no reason we'd have to wait).

> I wonder if it's not better to instead only run this for a specific tag pattern or branch, or maybe on every push to main?

See above, but can add the tag pattern if you agree (e.g. `v*`), but I think the rest of the triggers make sense as is (i.e. we probably don't want this workflow triggering on _any_ branch push, since pushes to a branch that isn't opened as a PR yet could be a little much).

Reply:
> We could do it for tags too (PR includes support for that), but we would have to add on.push.tags: [ v* ] which I have not included here (yet?).

Well, that was one of the things I was confused by, and why I was trying to understand your goal. I saw `if: startsWith(github.ref, 'refs/tags/v')` but wasn't sure how that would ever apply.

Reply:
Addressed via https://github.com/open-feature/dotnet-sdk/pull/173#discussion_r1459859465

tldr; removing references to release versions and tags from this PR to avoid any confusion